File: platforms/core-runtime/base-services/src/main/java/org/gradle/internal/buildtree/BuildModelParameters.java
/*
 * Copyright 2024 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gradle.internal.buildtree;

import org.gradle.internal.service.scopes.Scope;
import org.gradle.internal.service.scopes.ServiceScope;
import org.jspecify.annotations.Nullable;

import java.util.Map;

@ServiceScope(Scope.BuildTree.class)
public interface BuildModelParameters {

    /**
     * Whether project-scoped work should use project-lock or build-lock to synchronize,
     * allowing work from different projects to run in parallel when set to true.
     * <p>
     * Most notably, this allows parallel execution of tasks from different projects.
     * <p>
     * Note that this does not synchronize work from different builds.
     * <ul>
     * <li>Vintage: controlled by {@code --parallel} (or its property)
     * <li>CC: controlled by {@code --parallel} (or its property)
     * <li>IP: always enabled
     * </ul>
     */
    boolean isParallelProjectExecution();

    boolean isConfigureOnDemand();

    boolean isConfigurationCache();

    /**
     * User-friendly reason explaining why CC is disabled based on requested build options.
     * <p>
     * A common reason is the request for Gradle features incompatible with CC, such as {@code --export-keys}.
     */
    @Nullable
    String getConfigurationCacheDisabledReason();

    boolean isConfigurationCacheParallelStore();

    boolean isConfigurationCacheParallelLoad();

    boolean isIsolatedProjects();

    /**
     * Whether projects should be configured in parallel.
     * <p>
     * This should only take effect if {@link #isConfigureOnDemand() configure-on-demand}
     * is not making us skip eager project configuration.
     */
    boolean isParallelProjectConfiguration();

    /**
     * When {@link  #isIsolatedProjects()} is true, should project state be invalidated when a project it is coupled with changes?
     * This parameter is only used for benchmarking purposes.
     */
    boolean isInvalidateCoupledProjects();

    /**
     * Should model dependencies between projects be treated as project dependencies with respect to invalidation?
     * <p>
     * This parameter is only used for benchmarking purposes.
     */
    boolean isModelAsProjectDependency();

    /**
     * True when the build action requires to build Tooling Models.
     * <p>
     * When true, Gradle's "build model" such Gradle and Project state cannot be discarded
     * even after the tasks have been executed, because the Tooling Model Builders can run after tasks.
     */
    boolean isModelBuilding();

    /**
     * Determines whether nested build actions provided in {@code BuildController.run(actions)} can run in parallel.
     * <ul>
     * <li>Vintage: controlled by {@code --parallel}
     * <li>CC: not applicable, since CC is always disabled for model building invocations
     * <li>IP: always enabled
     * </ul>
     */
    boolean isParallelModelBuilding();

    /**
     * Determines whether models produced by tooling model builders are individually cached.
     * <p>
     * With IP, we can assume that for project-scoped models, their effective inputs are a subset of the Project state.
     * If the Project is up to date, we can serve its models from a cache instead of recomputing them on subsequent runs.
     * <p>
     * Always false for Vintage and CC.
     */
    boolean isCachingModelBuilding();

    /**
     * Returns true if the model building is resilient so some failures in model building.
     *
     * @return true if the model building is resilient, false otherwise
     */
    boolean isResilientModelBuilding();

    /**
     * Collects all properties and their values in a map for logging and testing purposes.
     */
    Map<String, ? extends @Nullable Object> toDisplayMap();
}

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/buildtree/control/BuildModelParameters.kt
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gradle.internal.buildtree.control

import org.gradle.internal.buildtree.BuildModelParameters


internal sealed class AbstractBuildModelParameters : BuildModelParameters {

    override fun toDisplayMap(): Map<String, Any?> = mapOf(
        "cachingModelBuilding" to isCachingModelBuilding,
        "configurationCache" to isConfigurationCache,
        "configurationCacheDisabledReason" to configurationCacheDisabledReason,
        "configurationCacheParallelLoad" to isConfigurationCacheParallelLoad,
        "configurationCacheParallelStore" to isConfigurationCacheParallelStore,
        "configureOnDemand" to isConfigureOnDemand,
        "invalidateCoupledProjects" to isInvalidateCoupledProjects,
        "isolatedProjects" to isIsolatedProjects,
        "modelAsProjectDependency" to isModelAsProjectDependency,
        "modelBuilding" to isModelBuilding,
        "parallelModelBuilding" to isParallelModelBuilding,
        "parallelProjectConfiguration" to isParallelProjectConfiguration,
        "parallelProjectExecution" to isParallelProjectExecution,
        "resilientModelBuilding" to isResilientModelBuilding,
    )
}

internal class GradleVintageMode(
    private val modelBuilding: Boolean,
    private val parallelProjectExecution: Boolean,
    private val configureOnDemand: Boolean,
    private val configurationCacheDisabledReason: String?,
    private val parallelModelBuilding: Boolean,
    private val resilientModelBuilding: Boolean,
) : AbstractBuildModelParameters() {

    override fun isParallelProjectExecution(): Boolean = parallelProjectExecution

    override fun isConfigureOnDemand(): Boolean = configureOnDemand

    override fun isConfigurationCache(): Boolean = false
    override fun getConfigurationCacheDisabledReason(): String? = configurationCacheDisabledReason
    override fun isConfigurationCacheParallelStore(): Boolean = false
    override fun isConfigurationCacheParallelLoad(): Boolean = false

    override fun isIsolatedProjects(): Boolean = false
    override fun isParallelProjectConfiguration(): Boolean = false
    override fun isInvalidateCoupledProjects(): Boolean = false
    override fun isModelAsProjectDependency(): Boolean = false

    override fun isModelBuilding(): Boolean = modelBuilding
    override fun isParallelModelBuilding(): Boolean = parallelModelBuilding
    override fun isCachingModelBuilding(): Boolean = false
    override fun isResilientModelBuilding(): Boolean = resilientModelBuilding

    override fun toString(): String = "GradleVintageMode"
}

internal class GradleConfigurationCacheMode(
    private val parallelProjectExecution: Boolean,
    private val configureOnDemand: Boolean,
    private val configurationCacheParallelStore: Boolean,
    private val configurationCacheParallelLoad: Boolean,
) : AbstractBuildModelParameters() {

    override fun isParallelProjectExecution(): Boolean = parallelProjectExecution

    override fun isConfigureOnDemand(): Boolean = configureOnDemand

    override fun isConfigurationCache(): Boolean = true
    override fun getConfigurationCacheDisabledReason(): String? = null
    override fun isConfigurationCacheParallelStore(): Boolean = configurationCacheParallelStore
    override fun isConfigurationCacheParallelLoad(): Boolean = configurationCacheParallelLoad

    override fun isIsolatedProjects(): Boolean = false
    override fun isParallelProjectConfiguration(): Boolean = false
    override fun isInvalidateCoupledProjects(): Boolean = false
    override fun isModelAsProjectDependency(): Boolean = false

    override fun isModelBuilding(): Boolean = false
    override fun isParallelModelBuilding(): Boolean = false
    override fun isCachingModelBuilding(): Boolean = false
    override fun isResilientModelBuilding(): Boolean = false

    override fun toString(): String = "GradleConfigurationCacheMode"
}

internal class GradleIsolatedProjectsMode(
    private val modelBuilding: Boolean,
    private val parallelProjectExecution: Boolean,
    private val configureOnDemand: Boolean,
    private val configurationCacheParallelStore: Boolean,
    private val parallelProjectConfiguration: Boolean,
    private val cachingModelBuilding: Boolean,
    private val parallelModelBuilding: Boolean,
    private val invalidateCoupledProjects: Boolean,
    private val modelAsProjectDependency: Boolean,
    private val resilientModelBuilding: Boolean
) : AbstractBuildModelParameters() {

    override fun isParallelProjectExecution(): Boolean = parallelProjectExecution

    override fun isConfigureOnDemand(): Boolean = configureOnDemand

    override fun isConfigurationCache(): Boolean = true
    override fun getConfigurationCacheDisabledReason(): String? = null
    override fun isConfigurationCacheParallelStore(): Boolean = configurationCacheParallelStore
    override fun isConfigurationCacheParallelLoad(): Boolean = true

    override fun isIsolatedProjects(): Boolean = true
    override fun isParallelProjectConfiguration(): Boolean = parallelProjectConfiguration
    override fun isInvalidateCoupledProjects(): Boolean = invalidateCoupledProjects
    override fun isModelAsProjectDependency(): Boolean = modelAsProjectDependency

    override fun isModelBuilding(): Boolean = modelBuilding
    override fun isParallelModelBuilding(): Boolean = parallelModelBuilding
    override fun isCachingModelBuilding(): Boolean = cachingModelBuilding
    override fun isResilientModelBuilding(): Boolean = resilientModelBuilding

    override fun toString(): String = "GradleIsolatedProjectsMode"
}

